### PR TITLE
p2p: fix DiscReason encoding/decoding

### PR DESCRIPTION
**Pull request by lorenzo-dev1**
This PR reverts back the change to the `DiscReason` rlp decoding done in https://github.com/ethereum/go-ethereum/pull/24507

It also adds a test to check for correctness. The current code in `master` will fail the test since the rlp decoding of DiscReason will fail. However the error returned from `rlp.Decode` is not checked, so the client proceeds using the default value for DiscReason, which is `DiscRequested`.

This PR also prevents a possible nil pointer dereference when closing an rlpx connection.

### CODE REVIEW

**Comment by lorenzo-dev1** [comment]
@fjl @zsfelfoldi I actually realized that the original decoding in the PR works with `uint8`. However then should the decoding with a struct be used also in `readProtocolHandshake` in `transport.go`. 

If so, I'll revert my changes and make the modification in `readProtocolHandshake`.

**Comment by lorenzo-dev1** [comment]
I noticed that the test `TestPeerDisconnect` will fail with this implementation, however I think the test might be wrong. Because it calls `SendItems` to send the disconnect message, which in turns calls `EncodeToReader`. 

However when closing a connection the client simply uses `rlp.Encode` to send the message. 

If my understanding is correct, I will fix that test too.

File: p2p/peer_test.go
Comment (holiman) [inline p2p/peer_test.go:25]: these should be added to the import-paragraph further down, not up here

**Comment by fjl** [comment]
Could you please provide some background why you are reverting https://github.com/ethereum/go-ethereum/pull/24507

**Comment by fjl** [comment]
Ahh I see why. So the server sends the reason as a byte array... In the spec it is defined to be an RLP list.

The fix needs to be applied on the writing side, i.e. in p2p/transport.go. The decoding part is actually correct. Though we could handle both cases as well.

**Comment by lorenzo-dev1** [comment]
> Ahh I see why. So the server sends the reason as a byte array... In the spec it is defined to be an RLP list.
> 
> The fix needs to be applied on the writing side, i.e. in p2p/transport.go. The decoding part is actually correct. Though we could handle both cases as well.

@fjl ah I see! I will implement the fix in the writing side today then!

**Comment by lorenzo-dev1** [comment]
@fjl should be good now, if you like the approach I will squash all commits

**Review by fjl — APPROVED** [review]
I have updated the change a bit. No need to squash since we use squash merge anyway.

File: p2p/peer_error.go
Comment (lorenzo-dev1) [inline p2p/peer_error.go:91]: @fjl  I think this bumps the length of the `discReasonToString` to 256... So it means that the `String()` func will never return the `uknown reason` error. it will return empty string instaed

Comment (fjl) [reply p2p/peer_error.go:91]: Good point, thanks!

**Comment by fjl** [comment]
About `nil` pointer isssue: I have removed the check for `t.conn == nil` since the `conn` field will never be nil.

**Review by NaNOteh — APPROVED** [review]

File: p2p/peer_test.go
Comment (lorenzo-dev1) [reply p2p/peer_test.go]: These changes are not present anymore

**Comment by lorenzo-dev1** [comment]
Seems like I cannot merge the PR even if approved, should I wait for other approvals?

**Comment by fjl** [comment]
It runs on AppVeyor. I will merge your PR later, still want to verify something about it.

**Pull request merged**
